tokio-trace-serde/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+use std::fmt;
+
+use serde::{
+    Serialize,
+    ser::{
+        SerializeMap,
+        SerializeStruct,
+        Serializer,
+    },
+};
+
+use tokio_trace::{
+    event::Event,
+    field::{Field, Visit},
+    metadata::Metadata,
+    span::{Attributes, Record},
+};
+
+#[derive(Debug)]
+pub struct FieldVisitor(pub (crate) Field);
+
+impl Serialize for FieldVisitor {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
+        let mut state = serializer.serialize_struct("Field", 1)?;
+        state.serialize_field("name", self.0.name())?;
+        state.end()
+    }
+}
+
+#[derive(Debug)]
+pub struct SerializeMetadata<'a>(pub (crate) &'a Metadata<'a>);
+
+impl<'a> Serialize for SerializeMetadata<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
+        let mut state = serializer.serialize_struct("Metadata", 2)?;
+        state.serialize_field("name", self.0.name)?;
+        state.serialize_field("target", self.0.target)?;
+        state.end()
+    }
+}
+
+/// A Serde visitor to pull `Event` data out of a serialized stream
+#[derive(Debug)]
+pub struct SerializeEvent<'a>(pub (crate) Event<'a>);
+
+impl<'a> Serialize for SerializeEvent<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
+        let mut state = serializer.serialize_struct("SerializeEvent", 2)?;
+        state.serialize_field("metadata", &SerializeMetadata(self.0.metadata()))?;
+        state.end()
+    }
+}
+
+/// A Serde visitor to pull `Attributes` data out of a serialized stream
+#[derive(Debug)]
+pub struct SerializeAttributes<'a>(pub (crate) Attributes<'a>);
+
+impl<'a> Serialize for SerializeAttributes<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
+        let mut state = serializer.serialize_struct("SerializeAttributes", 3)?;
+        state.serialize_field("metadata", &SerializeMetadata(self.0.metadata()))?;
+        if let Some(id) = self.0.parent() {
+            // TODO this probably isn't how we want it
+            state.serialize_field("parent", &format!("Explicit: {:?}", id))?;
+        }
+
+        else if self.0.is_root() {
+            state.serialize_field("parent", "Root")?;
+        }
+
+        else {
+            state.serialize_field("parent", "Current")?;
+        }
+
+        state.end()
+    }
+}
+
+/// A Serde visitor to pull `Record` data out of a serialized stream
+#[derive(Debug)]
+pub struct RecordValues<'a>(pub (crate) Record<'a>);
+
+impl<'a> Serialize for RecordValues<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
+        let our_state = serializer.serialize_map(Some(1))?;
```

Comment:
I believe the parameter passed to `serialize_map` is supposed to be the number of entries in the map. Since we don't know that up front, I think this should be:
```suggestion
        let our_state = serializer.serialize_map(None)?;
```

We should probably consider adding a `len` method to `tokio_trace::span::Record` upstream; once that is added we could change this to call `self.0.len()`.

Reply:
Also, TIOLI, but if you named this `let` binding `serializer`
```suggestion
        let serializer = serializer.serialize_map(Some(1))?;
```

you could pass it to the `SerdeVisitor` using the struct field initializer shorthand below, like this:
```rust
SerdeVisitor {
    serializer,
    state: Ok(()),
};
```